File: fortran/lax_solver.f90
module lax
    implicit none
    real    :: dt,delta
    real, dimension(:,:), allocatable :: F_x,F_y,c
    real, dimension(:,:), allocatable :: r_1,l_1,s_1,s_0
    integer :: Nx,Ny

contains

    subroutine initialise_lax()

        implicit none

        allocate(F_x(0:Nx+1,0:Ny+1),F_y(0:Nx+1,0:Ny+1),c(0:Nx+1,0:Ny+1))
        allocate(r_1(0:Nx+1,0:Ny+1),l_1(0:Nx+1,0:Ny+1))
        allocate(s_0(0:Nx+1,0:Ny+1),s_1(0:Nx+1,0:Ny+1))
        
        c(0:Nx+1,0:Ny+1) = 1.0

    end subroutine
    
    subroutine get_ghost(f_ghost,f)

        implicit none
        real, dimension(0:,0:) :: f_ghost
        real, dimension(:,:)   :: f
        integer :: ix,iy

        f_ghost(1:Nx,1:Ny) = f

        f_ghost(0,0)       = f(1,1)
        f_ghost(Nx+1,0)    = f(Nx,1)
        f_ghost(0,Ny+1)    = f(1,Ny)
        f_ghost(Nx+1,Ny+1) = f(Nx,Ny)

        ix = 0
        do iy = 1,Ny
            f_ghost(ix,iy) = f(1,iy)
        end do

        iy = 0
        do ix = 1,Nx
            f_ghost(ix,iy) = f(ix,1)
        end do

        ix = Nx+1
        do iy = 1,Ny
            f_ghost(ix,iy) = f(Nx,iy)
        end do

        iy = Ny+1
        do ix = 1,Nx
            f_ghost(ix,iy) = f(ix,Ny)
        end do

    end subroutine

    subroutine lax_update(f,jx,jy)

        implicit none
        real, dimension(0:,0:) :: f,jx,jy
        real, dimension(:,:), allocatable :: f_next
        integer :: ix,iy
        
        allocate(f_next(0:Nx+1,0:Ny+1))
        
        do ix = 1,Nx
            do iy = 1,Ny
                f_next(ix,iy) = 0.25 * (f(ix-1,iy) + f(ix+1,iy) + f(ix,iy-1) + f(ix,iy+1))
                f_next(ix,iy) = f_next(ix,iy) - 0.5 * (jx(ix+1,iy) - jx(ix-1,iy) + jy(ix,iy+1) - jy(ix,iy-1))
            end do
        end do
        
        f = f_next

    end subroutine lax_update

    subroutine lax_step(r,l,s,psi)

        implicit none
        real, dimension(:,:) :: r,l,s,psi
        integer :: ix,iy

        call get_ghost(s_0,s)
        call get_ghost(s_1,s)
        call get_ghost(r_1,r)
        call get_ghost(l_1,l)
        
        do ix = 0,Nx+1
            do iy = 0,Ny+1
                F_x(ix,iy) = -dt/delta*c(ix,iy)*r_1(ix,iy)
                F_y(ix,iy) = -dt/delta*c(ix,iy)*l_1(ix,iy)
            end do
        end do
        call lax_update(s_1,F_x,F_y)
        
        do ix = 0,Nx+1
            do iy = 0,Ny+1
                F_x(ix,iy) = -dt/delta*c(ix,iy)*s_0(ix,iy)
                F_y(ix,iy) = 0.0
            end do
        end do
        call lax_update(r_1,F_x,F_y)
        ! Reflective boundaries
        r_1(1,:)  = 0.0
        r_1(Nx,:) = 0.0

        do ix = 0,Nx+1
            do iy = 0,Ny+1
                F_x(ix,iy) = 0.0
                F_y(ix,iy) = -dt/delta*c(ix,iy)*s_0(ix,iy)
            end do
        end do
        call lax_update(l_1,F_x,F_y)
        ! Reflective boundaries
        l_1(:,1)  = 0.0
        l_1(:,Ny) = 0.0

        do ix = 1,Nx
            do iy = 1,Ny
                r(ix,iy) = r_1(ix,iy)
                l(ix,iy) = l_1(ix,iy)
                s(ix,iy) = s_1(ix,iy)
                psi(ix,iy) = psi(ix,iy) + 0.5 * dt * (s_0(ix,iy) + s_1(ix,iy))
            end do
        end do

    end subroutine lax_step

end module lax

File: JAX/lax_solver.py
import jax
import jax.numpy as jnp

def c_uniform(xx,yy):
    return jnp.ones_like(xx)

def init_gaussian(x,y,sig,cfunc):
    xx,yy = jnp.meshgrid(x,y,indexing='ij')
    dist = jnp.sqrt(xx**2+yy**2)
    c = cfunc(xx,yy)
    r = -c*xx*jnp.exp(-0.5*dist**2/sig**2)/sig**2
    l = -c*yy*jnp.exp(-0.5*dist**2/sig**2)/sig**2
    s = jnp.zeros_like(xx)
    psi = jnp.exp(-0.5*dist**2/sig**2)
    return (r,l,s,psi),c

def get_ghost(f):
    f_x     = jnp.concatenate((f[0:1,:],f,f[-1:,:]),axis=0)
    f_ghost = jnp.concatenate((f_x[:,0:1],f_x,f_x[:,-1:]),axis=1)
    return f_ghost

def Lax_Update(f,jx,jy):
    f_1 = 0.25 * (f[:-2,1:-1] + f[2:,1:-1] + f[1:-1,:-2] + f[1:-1,2:])
    f_1 -= 0.5 * (jx[2:,:] - jx[:-2,:] + jy[:,2:] - jy[:,:-2])
    return f_1

def generate_Lax_Step(args):
    dt = args['dt']
    delta = args['delta']
    c = args['c']
    Nx,Ny = args['Nx'],args['Ny']
    def Lax_Step(u):
        r_0,l_0,s_0,psi_0 = u

        # Reflective boundaries
        r_0_ghost = get_ghost(r_0)
        l_0_ghost = get_ghost(l_0)
        s_0_ghost = get_ghost(s_0)
        c_ghost   = get_ghost(c)

        Fr_x = -dt/delta*c_ghost[:,1:-1]*s_0_ghost[:,1:-1]
        Fr_y = jnp.zeros((Nx,Ny+2))
        r_1 = Lax_Update(r_0_ghost,Fr_x,Fr_y)
        # Reflective boundaries
        r_1 = r_1.at[0,:].set(0.0)
        r_1 = r_1.at[-1,:].set(0.0)

        Fl_x = jnp.zeros((Nx+2,Ny))
        Fl_y = -dt/delta*c_ghost[1:-1,:]*s_0_ghost[1:-1,:]
        l_1 = Lax_Update(l_0_ghost,Fl_x,Fl_y)
        # Reflective boundaries
        l_1 = l_1.at[:,0].set(0.0)
        l_1 = l_1.at[:,-1].set(0.0)

        Fs_x = -dt/delta*c_ghost[:,1:-1]*r_0_ghost[:,1:-1]
        Fs_y = -dt/delta*c_ghost[1:-1,:]*l_0_ghost[1:-1,:]
        s_1 = Lax_Update(s_0_ghost,Fs_x,Fs_y)

        psi_1 = psi_0 + 0.5 * dt * (s_0 + s_1)

        u = r_1,l_1,s_1,psi_1

        return u
    
    return jax.jit(Lax_Step)